File: packages/spooky-ts/src/services/event-system.ts
import { RecordId } from "@spooky/query-builder";

export type EventDefinition<T extends EventType, P> = {
  type: T;
  payload: P;
};

export const AuthEventTypes = {
  Authenticated: "AUTHENTICATED",
  Deauthenticated: "DEAUTHENTICATED",
} as const;

export type EventTypeMap = {
  [AuthEventTypes.Authenticated]: EventDefinition<
    typeof AuthEventTypes.Authenticated,
    {
      userId: RecordId;
      token: string;
    }
  >;
  [AuthEventTypes.Deauthenticated]: EventDefinition<
    typeof AuthEventTypes.Deauthenticated,
    never
  >;
};

export type Event<T extends EventType> = EventTypeMap[T];

export type EventType = keyof EventTypeMap;

export type EventHandler<T extends EventType> = (event: Event<T>) => void;

type InnerEventHandler<T extends EventType> = {
  id: number;
  handler: EventHandler<T>;
};

export class AuthManagerService {
  private subscriberId: number = 0;
  private isProcessing: boolean = false;
  private buffer: Event<EventType>[];
  private subscribers: {
    [K in EventType]: Map<number, InnerEventHandler<K>>;
  };
  private subscribersTypeMap: Map<number, EventType>;
  private lastEvents: {
    [K in EventType]?: Event<K>;
  };

  constructor() {
    this.buffer = [];
    this.subscribers = {
      [AuthEventTypes.Authenticated]: new Map(),
      [AuthEventTypes.Deauthenticated]: new Map(),
    };
    this.lastEvents = {};
    this.subscribersTypeMap = new Map();
  }

  subscribe<T extends EventType>(type: T, handler: EventHandler<T>): number {
    const id = this.subscriberId++;
    this.subscribers[type].set(id, {
      id,
      handler,
    });
    this.subscribersTypeMap.set(id, type);
    return id;
  }

  unsubscribe(id: number): boolean {
    const type = this.subscribersTypeMap.get(id);
    if (type) {
      this.subscribers[type].delete(id);
      this.subscribersTypeMap.delete(id);
      return true;
    }
    return false;
  }

  addEvent(event: Event<EventType>): void {
    this.buffer.push(event);
    this.scheduleProcessing();
  }

  private scheduleProcessing(): void {
    if (!this.isProcessing) {
      queueMicrotask(() => this.processEvents());
    }
  }

  private async processEvents(): Promise<void> {
    if (this.isProcessing) return;
    this.isProcessing = true;

    try {
      while (this.dequeue());
    } finally {
      this.isProcessing = false;
    }
  }

  private dequeue(): boolean {
    const event = this.buffer.shift();
    if (!event) return false;

    this.setLastEvent(event.type, event);
    this.broadcastEvent(event.type, event);
    return true;
  }

  private setLastEvent<T extends EventType>(type: T, event: Event<T>): void {
    this.lastEvents[type] = event;
  }

  private broadcastEvent<T extends EventType>(type: T, event: Event<T>): void {
    const subscribers = this.subscribers[type].values();
    for (const subscriber of subscribers) {
      subscriber.handler(event);
    }
  }
}

File: packages/spooky_core/lib/src/events/event_system.dart
import 'dart:async';

/// A single event: a string [type] and an arbitrary [payload].
///
/// The TS system is generic over an `EventTypeMap`; Dart keeps payloads
/// dynamic at the boundary and relies on typed facades / call sites for
/// safety. Mirrors the TS `Event<E, T>` shape (`{ type, payload }`).
class SpookyEvent {
  const SpookyEvent(this.type, this.payload);
  final String type;
  final dynamic payload;
}

typedef SpookyEventHandler = void Function(SpookyEvent event);

/// Debounce configuration for [EventSystem.addEvent] (TS `PushEventOptions`).
class PushEventOptions {
  const PushEventOptions({this.debounced});
  final DebouncedConfig? debounced;
}

class DebouncedConfig {
  const DebouncedConfig({required this.key, required this.delay});
  final String key;
  final int delay;
}

/// Subscription options (TS `EventSubscriptionOptions`).
class EventSubscriptionOptions {
  const EventSubscriptionOptions({this.immediately = false, this.once = false});
  final bool immediately;
  final bool once;
}

class _InnerHandler {
  _InnerHandler(this.id, this.handler, this.once);
  final int id;
  final SpookyEventHandler handler;
  final bool once;
}

/// Type-safe pub/sub with microtask-batched delivery, last-event replay, and
/// key/delay debounce. Faithful port of the TS `EventSystem`.
///
/// Kept as a callback registry (not Dart `Stream`s) to preserve numeric
/// subscription ids, the buffering/`lastEvents` semantics, and `once` /
/// `immediately` behavior the modules depend on.
class EventSystem {
  EventSystem(this._eventTypes) {
    for (final t in _eventTypes) {
      _subscribers[t] = <int, _InnerHandler>{};
    }
  }

  final List<String> _eventTypes;
  int _subscriberId = 0;
  bool _isProcessing = false;
  final List<SpookyEvent> _buffer = [];
  final Map<String, Map<int, _InnerHandler>> _subscribers = {};
  final Map<int, String> _subscribersTypeMap = {};
  final Map<String, SpookyEvent> _lastEvents = {};
  final Map<String, Timer> _debouncedEvents = {};

  List<String> get eventTypes => _eventTypes;

  /// Subscribe [handler] to [type]; returns an id for [unsubscribe].
  int subscribe(
    String type,
    SpookyEventHandler handler, [
    EventSubscriptionOptions options = const EventSubscriptionOptions(),
  ]) {
    final id = _subscriberId++;
    _subscribers[type]![id] = _InnerHandler(id, handler, options.once);
    _subscribersTypeMap[id] = type;
    if (options.immediately) {
      final last = _lastEvents[type];
      if (last != null) handler(last);
    }
    return id;
  }

  /// Subscribe one handler to several event types; returns all ids.
  List<int> subscribeMany(
    List<String> types,
    SpookyEventHandler handler, [
    EventSubscriptionOptions options = const EventSubscriptionOptions(),
  ]) =>
      types.map((t) => subscribe(t, handler, options)).toList();

  bool unsubscribe(int id) {
    final type = _subscribersTypeMap[id];
    if (type != null) {
      _subscribers[type]!.remove(id);
      _subscribersTypeMap.remove(id);
      return true;
    }
    return false;
  }

  /// Emit [type] with [payload].
  void emit(String type, dynamic payload) =>
      addEvent(SpookyEvent(type, payload));

  /// Add a fully-constructed event, optionally debounced by key.
  void addEvent(SpookyEvent event, [PushEventOptions? options]) {
    final debounced = options?.debounced;
    if (debounced != null) {
      _handleDebouncedEvent(event, debounced.key, debounced.delay);
      return;
    }
    _buffer.add(event);
    _scheduleProcessing();
  }

  void _handleDebouncedEvent(SpookyEvent event, String key, int delay) {
    _debouncedEvents[key]?.cancel();
    _debouncedEvents[key] = Timer(Duration(milliseconds: delay), () {
      _debouncedEvents.remove(key);
      _buffer.add(event);
      _scheduleProcessing();
    });
  }

  void _scheduleProcessing() {
    if (!_isProcessing) {
      scheduleMicrotask(_processEvents);
    }
  }

  void _processEvents() {
    if (_isProcessing) return;
    _isProcessing = true;
    try {
      while (_dequeue()) {}
    } finally {
      _isProcessing = false;
    }
  }

  bool _dequeue() {
    if (_buffer.isEmpty) return false;
    final event = _buffer.removeAt(0);
    _lastEvents[event.type] = event;
    _broadcastEvent(event.type, event);
    return true;
  }

  void _broadcastEvent(String type, SpookyEvent event) {
    // Copy to a list so `once` unsubscribes don't mutate during iteration.
    for (final subscriber in _subscribers[type]!.values.toList()) {
      subscriber.handler(event);
      if (subscriber.once) {
        unsubscribe(subscriber.id);
      }
    }
  }
}

EventSystem createEventSystem(List<String> eventTypes) =>
    EventSystem(eventTypes);
